app/scripts/modules/core/src/presentation/spel/SpelInput.tsx

```diff
@@ -0,0 +1,59 @@
+import * as React from 'react';
+import { SpelService } from './SpelService';
+
+import {
+  ITextInputProps,
+  TextAreaInput,
+  asyncMessage,
+  messageMessage,
+  useDebouncedValue,
+  useInternalValidator,
+  useIsMountedRef,
+  useData,
+  warningMessage,
+} from 'core/presentation';
+
+import './SpelInput.less';
+
+interface ISpelInputProps extends ITextInputProps {
+  previewExecutionId: string;
+  previewStageId: string;
+}
+
+/** An Input that previews Spel expressions */
+export function SpelInput(props: ISpelInputProps) {
+  const { previewExecutionId, previewStageId, ...rest } = props;
+  const [expression, isDebouncing] = useDebouncedValue(props.value, 300);
+  const isInitialRender = !useIsMountedRef().current;
+  const hasNewlines = !!/\n/.exec(rest.value);
+
+  function previewExpressionValidator() {
+    const { status, error, result } = fetchSpelPreview;
+    if (status === 'NONE') {
+      return null;
+    } else if (status === 'PENDING' || isDebouncing) {
+      return asyncMessage('previewing expression...');
+    } else if (status === 'RESOLVED') {
+      // Render as a code block
+      return messageMessage('Preview\n\n\n```\n' + JSON.stringify(result, null, 2) + '\n```');
+    } else if (status === 'REJECTED') {
+      return warningMessage(error);
+    }
+    return null;
+  }
+
+  useInternalValidator(props.validation, previewExpressionValidator);
+
+  const fetchSpelPreview = useData(
+    () => SpelService.evaluateExpression(expression, previewExecutionId, previewStageId),
+    null,
+    [expression, previewExecutionId, previewStageId],
+  );
+
+  // re-validate whenever an async event occurs
+  React.useEffect(() => {
+    !isInitialRender && props.validation && props.validation.revalidate && props.validation.revalidate();
```

Comment:
An aside, but the new version of ts has optional chaining which would be great for expressions like these.

Reply:
yes!  I can't wait to start using that.  Feel free to take a stab at updating our TS dependency.  I know there were some typing errors that were surfaced with typescript 3.5 (which is why we're on 3.4).

Reply:
Same! Maybe I will take a stab at it after I am done with this set of PRs